### PR TITLE
[Math] Fix bug in TFoam.

### PR DESCRIPTION
**Pull request by hageboeck**
[ROOT-10277] When creating a lot of datasets in RooFit, TFoam eventually
crashed, since it was using a TRefArray, which does some global UID
magic that overflows. Replacing the TRefArray by a simple vector of
indices fixes that.
This is possible, since the TRefArray only points to active elements in
another collection, so simply storing the index in that collection is
sufficient.

A ROOT-6.20 TFoam was created on lxplus, and written to a file. Reading
this file, and generating the 10 next random numbers is tested now.

@pcanal 
Please check the schema evolution for me. Is it correct to "read" an array of pointers like this?
```
  TFoamCell **fCells;
  fCells = onfile.fCells;
```
It seems to work, but I'm unsure if that's just by luck (e.g. since the file is still open).

### CODE REVIEW

**Comment by pcanal** [comment]
> Please check the schema evolution for me. Is it correct to "read" an array of pointers like this?

You may or may not have also do:
```
onfile.fCells = nullptr;
```
i.e. It is plausible that onfile.fCells is deleted after the I/O operation.

File: math/foam/inc/LinkDef.h [hunk 1]
```diff
@@ -10,5 +10,19 @@
 #pragma link C++ class TFoamCell+;
 #pragma link C++ class TFoam+;
 #pragma link C++ class TFoamSampler+;
-
+#pragma read sourceClass="TFoam" targetClass="TFoam" version="[1]" \
+  source="Int_t fNCells; TFoamCell **fCells; TRefArray *fCellsAct" target="fNCells,fCells,fCellsAct"\
+  include="TRefArray.h" \
+  code="{fNCells = onfile.fNCells; \
+         fCells = onfile.fCells; \
+         fCellsAct.clear(); \
```
Comment (hageboeck) [inline math/foam/inc/LinkDef.h:18]:
```suggestion
         onfile.fCells = nullptr; \
         fCellsAct.clear(); \
```

File: math/foam/inc/TFoam.h [hunk 2]
```diff
@@ -47,7 +46,7 @@ class TFoam : public TObject {
 
    TFoamMaxwt   *fMCMonit;    ///< Monitor of the MC weight for measuring MC efficiency
    Double_t   fMaxWtRej;      ///< Maximum weight in rejection for getting wt=1 events
-   TRefArray *fCellsAct;      ///< Array of pointers to active cells, constructed at the end of foam build-up
+   std::vector<Long_t> fCellsAct; ///< Index of active cells, constructed at the end of foam build-up
```
Comment (lmoneta) [inline math/foam/inc/TFoam.h:49]: Since now we are using an std::vector, should not we add an #include <vector>

Comment (hageboeck) [reply math/foam/inc/TFoam.h:50]: Oh, good idea, indeed.

**Pull request merged**